### PR TITLE
Add HW24.2 solution

### PR DESCRIPTION
**Pull request by OleksiiTromsa**

### CODE REVIEW

File: src/main/java/com/internet/shop/service/UserService.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+public interface UserService {
+    User create(User user);
+
+    User get(Long id);
+
+    List<User> getAll();
+
+    User update(User user);
+
+    Boolean delete(Long id);
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/service/UserService.java]: Why return wrapper here?

File: src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    @Inject
+    private OrderDao orderDao;
+    @Inject
+    private ShoppingCartDao shoppingCartDao;
+
+    @Override
+    public Order completeOrder(ShoppingCart shoppingCart) {
+        Order order = new Order(shoppingCart.getUserId());
+        order.setProducts(List.copyOf(shoppingCart.getProducts()));
+        shoppingCartDao.clear(shoppingCart);
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java]: We have to update ShoppingCart after changing

File: src/main/java/com/internet/shop/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+public class UserServiceImpl implements UserService {
+    @Inject
+    private UserDao userDao;
+    @Inject
+    private ShoppingCartDao shoppingCartDao;
+
+    @Override
+    public User create(User user) {
+        userDao.create(user);
+        shoppingCartDao.create(new ShoppingCart(user.getId()));
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/service/impl/UserServiceImpl.java]: According to comments to my HW we shouldn't do it (for example for admin users we don't need cart)

File: src/main/java/com/internet/shop/service/UserService.java (hunk 1 above)
Comment (OleksiiTromsa) [reply src/main/java/com/internet/shop/service/UserService.java]: Misprint, fixed

File: src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java (hunk 2 above)
Comment (OleksiiTromsa) [reply src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java]: Ok, I have refactored clearing logic

File: src/main/java/com/internet/shop/service/impl/UserServiceImpl.java (hunk 3 above)
Comment (OleksiiTromsa) [reply src/main/java/com/internet/shop/service/impl/UserServiceImpl.java]: Agree, didn't think about admin users. Refactored.

File: src/main/java/com/internet/shop/dao/OrderDao.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.internet.shop.dao;
+
+import com.internet.shop.model.Order;
+import java.util.List;
+import java.util.Optional;
+
+public interface OrderDao {
+    Order createOrder(Order order);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/OrderDao.java]:
```suggestion
    Order create(Order order);
```

File: src/main/java/com/internet/shop/dao/ShoppingCartDao.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.internet.shop.dao;
+
+import com.internet.shop.model.ShoppingCart;
+import java.util.Optional;
+
+public interface ShoppingCartDao {
+    ShoppingCart create(ShoppingCart shoppingCart);
+
+    Optional<ShoppingCart> getUserShoppingCard(Long userId);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/ShoppingCartDao.java]:
```suggestion
    Optional<ShoppingCart> getUserShoppingCart(Long userId);
```

File: src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    private OrderDao orderDao;
+    @Inject
+    private ShoppingCartDao shoppingCartDao;
+
+    @Override
+    public Order completeOrder(ShoppingCart shoppingCart) {
+        Order order = new Order(shoppingCart.getUserId());
+        order.setProducts(List.copyOf(shoppingCart.getProducts()));
+        shoppingCart.getProducts().clear();
+        shoppingCartDao.update(shoppingCart);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java]: давай використовувати для цього shoppingCartService.clear(shoppingCart);

File: src/main/java/com/internet/shop/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+@Service
+public class UserServiceImpl implements UserService {
+    @Inject
+    private UserDao userDao;
+
+    @Override
+    public User create(User user) {
+        userDao.create(user);
+        return user;
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/service/impl/UserServiceImpl.java]:
```suggestion
        return userDao.create(user);
```

File: src/main/java/com/internet/shop/dao/OrderDao.java (hunk 4 above)
Comment (OleksiiTromsa) [reply src/main/java/com/internet/shop/dao/OrderDao.java]: done

File: src/main/java/com/internet/shop/dao/ShoppingCartDao.java (hunk 5 above)
Comment (OleksiiTromsa) [reply src/main/java/com/internet/shop/dao/ShoppingCartDao.java]: fixed

File: src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java (hunk 6 above)
Comment (OleksiiTromsa) [reply src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java]: Done

File: src/main/java/com/internet/shop/service/impl/UserServiceImpl.java (hunk 7 above)
Comment (OleksiiTromsa) [reply src/main/java/com/internet/shop/service/impl/UserServiceImpl.java]: fixed

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**